### PR TITLE
Authentication + Sorting Spinners

### PR DESCRIPTION
**Pull request by BAFiogbe**
-- Added fireauth class: Used to authenticate new and existing users
-- Added spinners for sorting
--Updated AndroidManifest.xml with new classes

### CODE REVIEW

File: app/src/main/java/com/example/happymeals/Login.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+/** Initial UI layer for user login
+ * This class is a candidate to be the entry point of the application -- its implementation has yet
+ *  to be approved. It features a login screen and will authenticate users. Upon successful authentication,
+ *  users can then access all their persistent data.
+ * @author sruduke bfiogbe
+ *
+ */
+
+
+public class Login extends AppCompatActivity {
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/Login.java]: This should be refactored to LoginActivity

File: app/src/main/java/com/example/happymeals/Login.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+     * This is the function called whenever the LoginScreen is created -- in our
+     * case, this is on the launch of the app (if approved by client). If implemented,
+     * upon successful user authentication, it will send an intent to open an instance of
+     * the MainActivity. From there, users will be able to access all of their data -- ingredients,
+     * recipes, meal plans, etc.
+     * @param savedInstanceState The instance state to restore the activity to (if applicable) {@link Bundle}
+     */
+
+
+    private Button login, register;
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/Login.java]: These names are very ambiguous, consider changing them to something like loginButton, or forgotPasswordLabel, etc.

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -29,14 +36,19 @@
         </activity>
 
         <activity
-            android:name=".MainActivity"
+            android:name=".Login"
```
Comment (jeastgaard) [inline app/src/main/AndroidManifest.xml]: When name change happens for Login class, change the activity name here as well.

File: app/src/main/java/com/example/happymeals/Login.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+                //2 empty password passed
+
+                if( TextUtils.isEmpty(inputPass)) {
+                    password.setError("Password cannot be empty");
+                }
+                //3 Password length cannot be less than 8 chars
+
+                if( inputPass.length() < 8) {
+                    password.setError("Password must be greater than 8 characters");
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/Login.java]: Why are we telling the user how long the password must be on login? This is usually a registration thing.

File: app/src/main/java/com/example/happymeals/OutputListener.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.example.happymeals;
+
+public interface OutputListener {
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/OutputListener.java]: Is this output listener sitting in the main package? It should be moved to a package if so.  What is the purpose of this listener? Where should it be used? Need comments.

I suspect this listener should probably have a name change as the name "output"listener does not make sense to me. Can you justify this name?

File: app/src/main/java/com/example/happymeals/Register.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.TextView;
+
+import androidx.appcompat.app.AppCompatActivity;
+
+import com.example.happymeals.database.FireAuth;
+
+public class Register extends AppCompatActivity{
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/Register.java]: This should be renamed to RegisterActivity

File: app/src/main/java/com/example/happymeals/Register.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    private Button register;
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/Register.java]: Variable names are ambiguous.

File: app/src/main/java/com/example/happymeals/Register.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                }
+                fireAuth.getFireauth().userLogin(newUser, newPass, new OutputListener() {
+                    @Override
+                    public void onSuccess() {
+                        startActivity(new Intent(Register.this, MainActivity.class));
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        return;}
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/Register.java]: We required error logging on the failure of a firebase interaction.

File: app/src/main/java/com/example/happymeals/Register.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                    });
+
+                }
+            });
+            returnLogin.setOnClickListener(new View.OnClickListener(){
+                @Override
+                public void onClick(View v) {
+                    startActivity(new Intent(Register.this, Login.class));
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/Register.java]: You can't do this, it will create an infinite loop of activities being created i.e. login->register->login-> etc. all while the "paused" activities stay on the stack. You must finish this activity.

File: app/src/main/java/com/example/happymeals/Register.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                }
+                //3 Password length cannot be less than 8 chars
+
+                if (newPass.length() < 8) {
+                    password.setError("Password must be greater than 8 characters");
+                }
+                fireAuth.getFireauth().userLogin(newUser, newPass, new OutputListener() {
+                    @Override
+                    public void onSuccess() {
+                        startActivity(new Intent(Register.this, MainActivity.class));
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/Register.java]: You should return the success to the login activity, this login will be our main activity that is always present. The login activity should launch the main activity and this activity should finish.

File: app/src/main/java/com/example/happymeals/database/FireAuth.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+
+ */
+
+
+public class FireAuth {
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/database/FireAuth.java]: Don't need shorthand class names, should be FirebaseAuthenticationHandler, or just AuthenticationHandler/Manager/etc.

File: app/src/main/java/com/example/happymeals/database/FireAuth.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    public FirebaseFirestore fireStore;
+    public FirebaseAuth authenticate;
+    static private FireAuth fireauth;
+
+    // getter and setter for FireAuth
+    public static FireAuth getFireauth() {
+        // check if empty than initialize
+        if(fireauth == null) {
+            fireauth = new FireAuth();
+            fireauth.authenticate = FirebaseAuth.getInstance();
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/database/FireAuth.java]: Why is this not being done in the constructor?

File: app/src/main/java/com/example/happymeals/database/FireAuth.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+     */
+
+    public void userLogin(String user, String password, OutputListener listener){
+        fireauth.authenticate.signInWithEmailAndPassword(user, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
+            @Override
+            // test
+            public void onComplete(@NonNull Task<AuthResult> action) {
+                if (action.isSuccessful()) {
+                    listener.onSuccess();
+                    Log.d("TAG", "User has successfully authenticated");
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/database/FireAuth.java]: "TAG" needs to let the user know where this error message is coming from i.e. which component. TAG is not helpful in this sense. This goes for all the Log's in this class.

**Review by jeastgaard — CHANGES_REQUESTED** [review]
There are some name changes that need to take place and some comments with questions.

File: app/src/main/java/com/example/happymeals/OutputListener.java (hunk 5 above)
Comment (BAFiogbe) [reply app/src/main/java/com/example/happymeals/OutputListener.java]: it's an interface I use with custom methods to check if an output from another class is expected (successful) or fails. Honestly could be removed.

File: app/src/main/java/com/example/happymeals/database/FireAuth.java (hunk 12 above)
Comment (BAFiogbe) [reply app/src/main/java/com/example/happymeals/database/FireAuth.java]: It would do the same thing, but i'll rewrite it np.

File: app/src/main/java/com/example/happymeals/userlogin/LoginActivity.java [hunk 14]
```diff
@@ -100,7 +98,9 @@ public void onFailure(Exception e) {
         register.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View v) {
-                startActivity(new Intent(Login.this, Register.class));
+                startActivity(new Intent(LoginActivity.this, RegisterActivity.class));
+                // close this activity
+                finish();
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/userlogin/LoginActivity.java]: Have you debugged this line of code? Does this get called when you expect it to?

File: app/src/main/java/com/example/happymeals/userlogin/RegisterActivity.java [hunk 15]
```diff
@@ -1,23 +1,25 @@
...
+import com.example.happymeals.database.FirebaseAuthenticationHandler;
 
-public class Register extends AppCompatActivity{
-    private Button register;
+public class RegisterActivity extends AppCompatActivity{
+    private Button registerBtn;
     private TextView returnLogin;
-    private EditText user, password;
-    private FireAuth fireAuth;
+    private EditText userData, passData;
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/userlogin/RegisterActivity.java]: Let's have a talk about the naming of these. Data implies it holds information and is a data type. These are custom view fields in our .xml files. The name of the variables should reflect this.

File: app/src/main/java/com/example/happymeals/userlogin/RegisterActivity.java [hunk 16]
```diff
@@ -27,50 +29,54 @@ protected void onCreate( Bundle savedInstanceState ) {
...
 
                 }
             });
             returnLogin.setOnClickListener(new View.OnClickListener(){
                 @Override
                 public void onClick(View v) {
-                    startActivity(new Intent(Register.this, Login.class));
+                    startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
+                    // close this activity
+                    finish();
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/userlogin/RegisterActivity.java:79]: Same as above, curious if this works.

**Review by jeastgaard — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/happymeals/database/FirebaseAuthenticationHandler.java [hunk 17]
```diff
@@ -28,28 +28,20 @@ public class FirebaseAuthenticationHandler {
     // create FireAuth to function statically and allow login and register to access it
     public FirebaseFirestore fireStore;
     public FirebaseAuth authenticate;
-    static private FirebaseAuthenticationHandler fireAuth;
+    static private FirebaseAuthenticationHandler fireAuth = null;
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/database/FirebaseAuthenticationHandler.java]: Why did you remove this from the constructor???

File: app/src/main/java/com/example/happymeals/userlogin/LoginActivity.java [hunk 18]
```diff
@@ -38,7 +38,7 @@ public class LoginActivity extends AppCompatActivity {
     private Button login, register;
     private TextView forgotPassword;
     private EditText user, password;
-    private FirebaseAuthenticationHandler fireAuth;
+    private FirebaseAuthenticationHandler fireAuth = FirebaseAuthenticationHandler.getFireAuth();
```
Comment (jeastgaard) [inline app/src/main/java/com/example/happymeals/userlogin/LoginActivity.java]: Why is this outside of a method/function scope???

**Review by jeastgaard — APPROVED** [review]
